### PR TITLE
Add Multiple Round Robin With Adjusted Winner

### LINKED ISSUE #142: Multiple Round Robin with Adjusted Winners (MRR)

### PR DESCRIPTION
**Pull request by ketphan02**
### Sample run
```text
Class size: 20
-----
Team 2: 6 2 14 9 
-----
Team 4: 4 3 0 17 
-----
Team 1: 8 16 11 
-----
Team 6: 19 5 
-----
Team 5: 15 10 13 
-----
Team 3: 18 7 1 12 
```

Closes #142

### CODE REVIEW

File: api/ai/new/interfaces/algorithm_options.py [hunk 1]
```diff
@@ -72,3 +72,9 @@ def validate(self):
 class PathGaspAlgorithmOptions(AlgorithmOptions):
     def validate(self):
         super().validate()
+
+
+@dataclass
+class MultipleRoundRobinAlgorithmOptions(AlgorithmOptions):
```
Comment (Opeyem1a) [inline api/ai/new/interfaces/algorithm_options.py:78]: new algorithms will need to be added AlgorithmType enum

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: api/ai/new/multiple_round_robin_with_adjusted_winner/utils.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+def calculate_value(student: Student, requirements: List[ProjectRequirement]) -> int:
+    return sum(
+        [
+            1 if student.meets_requirement(requirement) else -1
+            for requirement in requirements
+        ]
+    )
+
+
+def is_oef1(team_i: "TeamWithValues", team_j: "TeamWithValues") -> bool:
```
Comment (Opeyem1a) [inline api/ai/new/multiple_round_robin_with_adjusted_winner/utils.py:16]: nit:
this function name is gross lmao

File: api/ai/new/multiple_round_robin_with_adjusted_winner/multiple_round_robin.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        self.projects = projects
+
+    @staticmethod
+    def _get_values_heap(
+        students: List[Student], teams: List[TeamWithValues]
+    ) -> Dict[int, List[StudentProjectValue]]:
+        """
+        Calculate each student values to each project and push them into a heap
+
+        This run in O(N * H)
```
Comment (Opeyem1a) [inline api/ai/new/multiple_round_robin_with_adjusted_winner/multiple_round_robin.py:66]: A comment from the other PR is why do we need to annotate the big O everywhere? Like what does it add to the codebase to have this as a comment in the code?

File: api/ai/new/multiple_round_robin_with_adjusted_winner/multiple_round_robin.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        }
+        for team in teams:
+            project = team.project
+            for student in students:
+                new_value = StudentProjectValue(student, project)
+                heappush(values_heap[project.id], new_value)
+
+        return values_heap
+
+    @staticmethod
```
Comment (Opeyem1a) [inline api/ai/new/multiple_round_robin_with_adjusted_winner/multiple_round_robin.py:80]: why make these methods static if you're just calling them with `self.` anyway?

If they are truly non-dependant on the `self.` then why aren't they utils?

(note: not every warning from pycharm means you need to fix it )

File: api/ai/new/multiple_round_robin_with_adjusted_winner/multiple_round_robin.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+from api.models.student import Student
+from api.models.team_set import TeamSet
+
+
+class MultipleRoundRobinWithAdjustedWinnerAlgorithm(Algorithm):
+    def __init__(
+        self,
+        algorithm_options: MultipleRoundRobinAlgorithmOptions,
+        team_generation_options: TeamGenerationOptions,
+        projects: List[Project],
```
Comment (Opeyem1a) [inline api/ai/new/multiple_round_robin_with_adjusted_winner/multiple_round_robin.py:54]: don't do this, if it's an argument of the algorithm then it should be in the `MultipleRoundRobinAlgorithmOptions`

Use the same casting technique from the other algorithms to cast it into the type you want

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: api/ai/new/multiple_round_robin_with_adjusted_winner/multiple_round_robin.py (hunk 3 above)
Comment (ketphan02) [reply api/ai/new/multiple_round_robin_with_adjusted_winner/multiple_round_robin.py:64]: It does not do much for the codebase, however, this will help a lot during the algorithm analysis process when formal documenting. If this is not preferred, I will add a TODO to remove it after formal documentation is established.

Comment (Opeyem1a) [reply api/ai/new/multiple_round_robin_with_adjusted_winner/multiple_round_robin.py:64]: Make an issue to remove them later, but I'm cool with this line of logic for now.
tldr; it provides you value personally for now, so it can stay, but long term (after it is formally documented) I think it's pretty low value forother devs working on this project, so can be removed

File: api/ai/new/interfaces/algorithm_options.py [hunk 6]
```diff
@@ -72,3 +73,11 @@ def validate(self):
...
+    projects: List[Project]
+
+    def validate(self):
+        super().validate()
```
Comment (Opeyem1a) [inline api/ai/new/interfaces/algorithm_options.py:83]: you can do this as another ticket, but are there no constraints on the list of projects? like if it's empty does the algorithm run fine? if not you need to add those validations here

**Review by Opeyem1a — APPROVED** [review]
you made it soldier
